### PR TITLE
[SYCL] Fix SYCL/Plugin/level_zero_sub_sub_device test case

### PR DESCRIPTION
**Pull request by uditagarwal97**
This test case was failing for a very long time. I investigated the failure and accordingly made the following changes:
1. Fixed RUN command by adding 'ZEX_NUMBER_OF_CCS=0:4' environment variable. This enables splitting sub-device into four CCS devices.
2. While partitioning sub-device into CCS, I changed 'partition_by_affinity_domain' to 'partition_by_cslice', in accordance to [#7676](https://github.com/intel/llvm/pull/7626).
3.  Reorganized the test case a bit to make it more "aesthetically appealing".

### CODE REVIEW

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 1]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
 #define INTER_NUM (150)
 #define KERNEL_NUM (2000)
 
-void run(std::vector<queue> &queues) {
+void make_queue_and_run_workload(std::vector<device> &subdevices) {
+
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]: nit, but I personally dislike empty line at the beginning of a function.
```suggestion
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 2]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+  std::cout << "[important] create " << subdevices.size()
+            << " sycl queues, one for each sub-sub device" << std::endl;
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]: This is kind of misleading. How is `subdevice.size` related to sub-sub-devices?

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 3]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
-  float *buffer_device0 = malloc_device<float>(N, queues[0]);
-
-  float *buffer_host1 = malloc_host<float>(N, queues[1]);
-  float *buffer_device1 = malloc_device<float>(N, queues[1]);
-
-  float *buffer_host2 = malloc_host<float>(N, queues[2]);
-  float *buffer_device2 = malloc_device<float>(N, queues[2]);
+  std::vector<queue> queues;
+  std::vector<float *> buffer_hosts;
+  std::vector<float *> buffer_devices;
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
  std::vector<float *> host_mem_ptrs;
  std::vector<float *> device_mem_ptrs;
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 4]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+
+  // Create queues for each subdevice
+  for (auto subdevice : subdevices) {
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
  for (auto &subdevice : subdevices) {
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 5]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+    queue temp(subdevice, {property::queue::enable_profiling(),
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
    queue q(subdevice, {property::queue::enable_profiling(),
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 6]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+                           property::queue::in_order()});
+    float *buffer_host = malloc_host<float>(N, temp);
+    float *buffer_device = malloc_device<float>(N, temp);
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]: The type is obvious from the template param, no need to spell it twice.

```suggestion
    auto *buffer_host = malloc_host<float>(N, temp);
    auto *buffer_device = malloc_device<float>(N, temp);
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 7]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
-  float *buffer_host1 = malloc_host<float>(N, queues[1]);
-  float *buffer_device1 = malloc_device<float>(N, queues[1]);
-
-  float *buffer_host2 = malloc_host<float>(N, queues[2]);
-  float *buffer_device2 = malloc_device<float>(N, queues[2]);
+  std::vector<queue> queues;
+  std::vector<float *> buffer_hosts;
+  std::vector<float *> buffer_devices;
+
+  // Create queues for each subdevice
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
  // Create queues for each subdevice.
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 8]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+    buffer_devices.push_back(buffer_device);
+    queues.push_back(temp);
   }
 
-  queues[0].memcpy(buffer_device0, buffer_host0, N * sizeof(float)).wait();
-  queues[1].memcpy(buffer_device1, buffer_host1, N * sizeof(float)).wait();
-  queues[2].memcpy(buffer_device2, buffer_host2, N * sizeof(float)).wait();
-  queues[3].memcpy(buffer_device3, buffer_host3, N * sizeof(float)).wait();
-
+  // Run workload
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
  // Run workload.
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 9]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
-            [=](nd_item<1> item) {
-              int i = item.get_global_linear_id();
-              buffer_device2[i] = buffer_device2[i] + float(2.0);
-            });
-      });
-      auto event3 = queues[3].submit([&](handler &h) {
-        h.parallel_for<class kernel3>(
+      for (int j = 0; j < queues.size(); j++) {
+
+        queue current_queue = queues[j];
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
      for (int j = 0; j < queues.size(); j++) {
        queue current_queue = queues[j];
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 10]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+        float *buffer_device = buffer_devices[j];
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]:
```suggestion
        float *device_mem = buffer_devices[j];
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 11]
```diff
@@ -25,89 +25,65 @@ using namespace std::chrono;
...
+      q.wait();
+  }
 
-  free(buffer_host3, queues[3]);
-  free(buffer_device3, queues[3]);
+  // De-allocate buffer hosts and devices
+  for (int j = 0; j < queues.size(); j++) {
+    sycl::free(buffer_devices[j], queues[j]);
+    sycl::free(buffer_hosts[j], queues[j]);
+  }
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]: The code is trivial, no need to state the same in a comment.
```suggestion
  for (int j = 0; j < queues.size(); j++) {
    sycl::free(buffer_devices[j], queues[j]);
    sycl::free(buffer_hosts[j], queues[j]);
  }
```

File: SYCL/Plugin/level_zero_sub_sub_device.cpp [hunk 12]
```diff
@@ -121,15 +97,15 @@ int main(void) {
 
   // watch out device here
   auto subdevices =
-      devices[1]
+      devices[0]
```
Comment (aelovikov-intel) [inline SYCL/Plugin/level_zero_sub_sub_device.cpp]: I'd probably change line 95 to

```
  device d;
```

and then just use `d` here, with lines 96-98 removed.

File: SYCL/Plugin/level_zero_sub_sub_device.cpp (hunk 2 above)
Comment (uditagarwal97) [reply SYCL/Plugin/level_zero_sub_sub_device.cpp]: Agree. I've updated the variable name to reflect the same.

**Review by aelovikov-intel — APPROVED** [review]

**Review by againull — APPROVED** [review]

**Pull request merged**